docs/proposals/THV-2106-virtual-mcp-server.md

```diff
@@ -0,0 +1,1035 @@
+5. **Authentication**: Extend existing auth middleware patterns
+6. **Elicitation**: Implement MCP elicitation protocol for composite tool user interaction
+
+### Token Exchange Integration (References THV-2063)
+
+Virtual MCP builds on the token exchange middleware from THV-2063-token-exchange-middleware.md:
+
+**Integration Points:**
+- Reuses `pkg/auth/tokenexchange` implementation for RFC 8693 token exchange
+- Per-backend exchange using separate `ExchangeConfig` instances
+- Token caching via pluggable `TokenCache` interface
+- Subject token extracted from incoming `Authorization` header
+
+**Discovery Pattern (Kubernetes Only):**
+
+Virtual MCP discovers backend authentication configs from `MCPExternalAuthConfig` resources referenced by each backend's `MCPServer.spec.externalAuthConfigRef` (see Kubernetes Configuration section for details). This leverages the existing controller pattern from `mcpexternalauthconfig_controller.go`.
+
+**Unified Configuration Model:**
+
+The platform-agnostic configuration model enables code reuse:
+- **CLI Adapter**: Loads from YAML files with environment variable substitution
+- **Kubernetes Adapter**: Loads from CRDs with automatic backend discovery
+- **Core Logic**: Both adapters convert to the same internal model
+
+This adapter pattern provides:
+- Consistent behavior across CLI and Kubernetes
+- Single test suite for business logic
+- Easy addition of new platforms (Docker Swarm, Nomad, etc.)
+
+### Composite Tool State Persistence
+
+Composite tools with elicitation require state persistence to handle long-running workflows that span multiple user interactions.
+
+**Phase 2 Approach (In-Memory Storage):**
+- Virtual MCP maintains workflow execution state in memory
+- State includes: current step, completed steps, elicitation responses, template variables
+- Clean storage interface enables future migration to persistent backends
+
+**Workflow Lifecycle:**
+- Each composite tool invocation receives a unique workflow ID
+- Server checkpoints state after each step completion and elicitation interaction
+- Default workflow timeout: 30 minutes
+- After timeout: automatic cleanup and pending elicitations treated as cancelled
+
+**Design Rationale:**
+This approach keeps implementation simple for Phase 2 while establishing patterns that can scale to distributed persistence (Redis, database) in Phase 3.
+
+## Alternative Approaches Considered
+
+1. **Kubernetes CRD**: More complex, requires operator changes
+2. **Standalone Service**: Loses integration with ToolHive infrastructure
+3. **LLM-generated backend**: Adds more randomness to the equation which is undesirable for agents.
+
+## Open Questions
+
+1. **Streaming responses across multiple backends** - How to merge/multiplex streaming responses from multiple backends for composite tools? Deferred to Phase 3.
+
+2. **Backend capability caching** - Should we cache backend capabilities (tools/resources/prompts) or query dynamically on each request?
```

Comment:
There maybe some cache hit/miss things here to get the best of both worlds but it depends on many things so I would defer this to later down the line